### PR TITLE
API review: add interception for DbCommand creation

### LINKED ISSUE #16715: Add interception for command creation

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #16715

### CODE REVIEW

**Comment by ajcvickers** [comment]
@roji @smitpatel Could one of you review this since there are not many people not on vacation right now. :-)

File: src/EFCore.Relational/Diagnostics/CommandCorrelatedEventData.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        /// </summary>
+        public virtual Guid CommandId { get; }
+
+        /// <summary>
+        ///     A correlation ID that identifies the <see cref="DbConnection" /> instance being used.
+        /// </summary>
+        public virtual Guid ConnectionId { get; }
+
+        /// <summary>
+        ///     The <see cref="DbCommand" /> method.
```
Comment (smitpatel) [inline src/EFCore.Relational/Diagnostics/CommandCorrelatedEventData.cs]: cref DbCommand or DbCommandMethod?

File: src/EFCore.Relational/Storage/RelationalCommand.cs [hunk 2]
```diff
@@ -450,11 +463,11 @@ public virtual RelationalDataReader ExecuteReader(RelationalCommandParameterObje
         {
             var (connection, context, logger) = (parameterObject.Connection, parameterObject.Context, parameterObject.Logger);
 
-            var command = CreateCommand(connection, parameterObject.ParameterValues);
+            var commandId = Guid.NewGuid();
+            var command = CreateCommand(parameterObject, parameterObject.ParameterValues, commandId, DbCommandMethod.ExecuteReader);
```
Comment (smitpatel) [inline src/EFCore.Relational/Storage/RelationalCommand.cs]: Given we are passing parameterObject itself, do we need to pass ParameterValues separately ?

File: src/EFCore.Relational/Storage/RelationalCommand.cs [hunk 3]
```diff
@@ -537,16 +553,34 @@ public virtual RelationalDataReader ExecuteReader(RelationalCommandParameterObje
...
+            Guid commandId,
+            DbCommandMethod commandMethod)
         {
-            Check.NotNull(connection, nameof(connection));
+            var (connection, context, logger) = (parameterObject.Connection, parameterObject.Context, parameterObject.Logger);
+            var connectionId = connection.ConnectionId;
 
-            var command = connection.DbConnection.CreateCommand();
+            var startTime = DateTimeOffset.UtcNow;
+            var stopwatch = Stopwatch.StartNew();
```
[the remark this answers is not in the document]
Comment (ajcvickers) [reply src/EFCore.Relational/Storage/RelationalCommand.cs]: Last I was aware of, activities don't work for interleaved events. (That is, when, for example, multiple commands are created before any of them finish.)

I added a note to #16197 about perf.

[the remark this answers is not in the document]
Comment (roji) [reply src/EFCore.Relational/Storage/RelationalCommand.cs]: I'm no expert but I seem to remember that interleaved activities are possible, at least with EventSource - an ActivityID is used to correlate between start and stop (we'd use the GUID here). [See this post](https://blogs.msdn.microsoft.com/vancem/2015/09/14/exploring-eventsource-activity-correlation-and-causation-features/).

**Pull request merged**